### ISSUE TITLE
Issue with rendering backup fonts

### ISSUE THREAD

**Issue by huntergoldstein** [jaagr/polybar#374]
# Config
```dosini
[bar/test]
monitor = ${env:MONITOR:eDP1}
width = 25%
modules-center = test
font-0 = Misc Termsyn;0
font-1 = Noto Sans CJK SC;0

[module/test]
type = custom/text
content = English; 한국어; English
```
# Expected Behavior
![2017-01-25-002223_479x27_scrot](https://cloud.githubusercontent.com/assets/5455867/22278836/6343ba44-e294-11e6-90de-b709489d5ed6.png)
## Info
```bash
$ polybar -vv
polybar 2.5.2

Features: +alsa +curl +i3 +mpd +network

X extensions: +xrandr (+monitors) -xrender -xdamage -xsync -xcomposite +xkb

Build type: RelWithDebInfo
Compiler: /usr/bin/c++
Compiler flags: -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -Wall -Wextra -Werror -O2 -pedantic -pedantic-errors
Linker flags: -Wl,-O1,--sort-common,--as-needed,-z,relro

ALSA_SOUNDCARD              default
BSPWM_SOCKET_PATH           /tmp/bspwm_0_0-socket
BSPWM_STATUS_PREFIX         W
BUILDER_SPACE_TOKEN         %__
CONNECTION_TEST_IP          8.8.8.8
PATH_ADAPTER                /sys/class/power_supply/%adapter%
PATH_BACKLIGHT_MAX          /sys/class/backlight/%card%/max_brightness
PATH_BACKLIGHT_VAL          /sys/class/backlight/%card%/brightness
PATH_BATTERY                /sys/class/power_supply/%battery%
PATH_CPU_INFO               /proc/stat
PATH_MEMORY_INFO            /proc/meminfo
PATH_TEMPERATURE_INFO       /sys/class/thermal/thermal_zone%zone%/temp

```
# Current Behavior
![2017-01-25-001440_480x22_scrot](https://cloud.githubusercontent.com/assets/5455867/22278706/8bd78482-e293-11e6-9d2d-84e8f13294fd.png)
## Info
```bash
$ polybar -vv
polybar 3.0.0-alpha

Features: +alsa +curl +i3 +mpd +network

X extensions: +randr (+monitors) -render -damage -sync -composite +xkb +xrm

Build type: RelWithDebInfo
Compiler: /usr/bin/c++
Compiler flags:  -Wall -Wextra -Werror -O2 -pedantic -pedantic-errors
Linker flags: 

ALSA_SOUNDCARD              default
BSPWM_SOCKET_PATH           /tmp/bspwm_0_0-socket
BSPWM_STATUS_PREFIX         W
BUILDER_SPACE_TOKEN         %__
CONNECTION_TEST_IP          8.8.8.8
PATH_ADAPTER                /sys/class/power_supply/%adapter%
PATH_BACKLIGHT_MAX          /sys/class/backlight/%card%/max_brightness
PATH_BACKLIGHT_VAL          /sys/class/backlight/%card%/brightness
PATH_BATTERY                /sys/class/power_supply/%battery%
PATH_CPU_INFO               /proc/stat
PATH_MEMORY_INFO            /proc/meminfo
PATH_TEMPERATURE_INFO       /sys/class/thermal/thermal_zone%zone%/temp
```
# Additional Comments
I'm using the `polybar` AUR package. I discovered the bug while using the `polybar-git` package; in practice the main issue was the font not changing back from Noto Sans to Termsyn. When I tried to isolate the issue (see attached conf) it got weirder.

**Comment by jaagr** [comment]
Hi and thanks for the detailed report.

Could you see if it works when manually resetting the font index:

~~~ dosini
content = "English; %{T2}한국어%{T-}; English"
~~~

**Comment by skystrife** [comment]
I can reproduce, and manually resetting the font index does indeed fix it for me. (I don't have the weird misalignment issue, though, just the font-not-resetting issue.)

It looks like the old `renderer::draw_textstring()` method would attempt to determine the best font for each glyph before drawing it. The new implementation in `context::operator<<(const textblock&)` instead tries to be somewhat greedy: as soon as it finds a font that matches a glyph, it attempts to determine how many subsequent characters it can also match and then renders all of them at once. The bug described here happens because "한국어; English" can all be matched by the fallback font, so it never attempts to switch back to `font-0` after it falls back to `font-1` on encountering the Korean text.

**Comment by jaagr** [comment]
This should be fixed as of 14fb86e. Now, any character that can't be rendered using the default/preferred font gets tested one by one against the remaining fonts. Once a match has been found, the matching returns to the top of the font list again.

**Comment by huntergoldstein** [comment]
Fixed for me on `master`. I cannot reproduce the misalignment anymore.
